### PR TITLE
All records implementation complete for all dictionary structures

### PR DESCRIPTION
**Pull request by danaack**
The corresponding unit tests have been updated and are successful.

### CODE REVIEW

File: src/dictionary/flat_file/flat_file_dictionary_handler.c [hunk 1]
```diff
@@ -197,6 +197,21 @@ ffdict_find(
 			}
 		}
 
+		case predicate_all_records: {
+			ffdict_cursor_t *ffdict_cursor = (ffdict_cursor_t *) (*cursor);
+
+			if (NULL == ((ff_file_t *) dictionary->instance)->start_of_data) {
```
Comment (Stickerpants) [inline src/dictionary/flat_file/flat_file_dictionary_handler.c]: I don't think this check is necessary

File: src/dictionary/flat_file/flat_file_dictionary_handler.c [hunk 2]
```diff
@@ -197,6 +197,21 @@ ffdict_find(
...
+				(*cursor)->status = cs_cursor_uninitialized;
+			}
+			else {
+				ffdict_cursor->current	= ((ff_file_t *) dictionary->instance)->start_of_data;
```
Comment (Stickerpants) [inline src/dictionary/flat_file/flat_file_dictionary_handler.c]: Here we should call `ffdict_scan` like we do on line 189 and behave accordingly. This should let us "exit early" in the case that there is no data to be read.

File: src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c [hunk 3]
```diff
@@ -182,27 +182,38 @@ oafdict_find(
...
+			break;
+		}
 
-				/* the cursor is ready to be consumed */
-				oadict_cursor->first	= location;
-				oadict_cursor->current	= location;
-				return err_ok;
+		case predicate_all_records: {
+			oafdict_cursor_t	*oafdict_cursor = (oafdict_cursor_t *) (*cursor);
+			file_hashmap_t		*hash_map		= ((file_hashmap_t *) dictionary->instance);
```
Comment (Stickerpants) [inline src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c]: possible to eliminate the duplicated code here?

File: src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c [hunk 4]
```diff
@@ -182,27 +182,38 @@ oafdict_find(
...
-				(*cursor)->status = cs_end_of_results;
-				return err_ok;
+			(*cursor)->status		= cs_cursor_initialized;
+			oafdict_cursor->first	= (hash_map->map_size) - 1;
+			oafdict_cursor->current = -1;
+
+			err_t err = oafdict_scan(oafdict_cursor);
+
+			if (cs_valid_data != err) {
+				(*cursor)->status = cs_cursor_uninitialized;
```
Comment (Stickerpants) [inline src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c]: If this signifies that there is no valid data in hashmap, we should return `cs_cursor_end_of_results`

File: src/dictionary/skip_list/skip_list_handler.c [hunk 5]
```diff
@@ -230,6 +232,21 @@ sldict_find(
 			break;
 		}
 
+		case predicate_all_records: {
+			sldict_cursor_t *sl_cursor = (sldict_cursor_t *) (*cursor);
+
+			if (NULL == skip_list->head->next[0]) {
+				(*cursor)->status = cs_cursor_uninitialized;
```
Comment (Stickerpants) [inline src/dictionary/skip_list/skip_list_handler.c]: `cs_cursor_end_of_results` here?

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 6]
```diff
@@ -324,20 +378,43 @@ test_cpp_wrapper_equality(
...
-		PLANCK_UNIT_ASSERT_TRUE(tc, eq_cursor->hasNext());
-		eq_cursor->next();
-		record_value++;
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, eq_key * 2, eq_cursor->getValue());
+		status = eq_cursor->next();
+		records_found++;
 	}
 
-	PLANCK_UNIT_ASSERT_TRUE(tc, !eq_cursor->hasNext());
+	PLANCK_UNIT_ASSERT_FALSE(tc, eq_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: nice :+1:

**Comment by Stickerpants** [comment]
LGTM

**Comment by graemedouglas** [comment]
LGTM

**Comment by graemedouglas** [comment]
LGTM

**Pull request merged**